### PR TITLE
WIP: Implement drag and drop for dashboard blocks

### PR DESCRIPTION
**Pull request by yanniboi**
Updated contacts tab to take blocks with regions and started endpoint to update block position.

### CODE REVIEW

File: config/install/contacts.contact_tab.summary.yml [hunk 1]
```diff
@@ -6,10 +6,13 @@ label: Summary
...
+blocks:
+  contact_summary_tab:
+    id: contact_summary_tab
+    label: ''
+    provider: contacts
+    label_display: '0'
+    region: left
+    weight: 0
+    context_mapping:
+      user: user
```
Comment (andrewbelcher) [inline config/install/contacts.contact_tab.summary.yml]: This should probably use the three blocks directly, Indiv/Org on left, Notes on right. We can then remove the summary tab block/theme/templates etc and be rid of them!

File: contacts.module [hunk 2]
```diff
@@ -26,6 +26,10 @@ function contacts_theme() {
       'variables' => ['tabs' => []],
       'template' => 'contacts-dash-tabs',
     ],
+    'contacts_dash_tab_content' => [
+      'variables' => ['content' => []],
+      'template' => 'contacts-dash-tab-content',
+    ],
```
Comment (andrewbelcher) [inline contacts.module]: Should this just be `contact_tab` for the theme function?

File: contacts.routing.yml [hunk 3]
```diff
@@ -28,3 +28,15 @@ contacts.add_org_form:
     _permission: 'add contacts'
   options:
     _admin_route: TRUE
+
+contacts.dash_block_move:
```
Comment (andrewbelcher) [inline contacts.routing.yml]: I wonder if it's worth putting this as a route under the contact_tab entity, so something like `entity.contact_tab.position_block`? Wonder if that would be more intuitive? I def think position block makes more sense than move block.

I also wonder whether we need to have a single POST endpoint that updates all the regions/weights for blocks on a single tab. If you re-arrange the blocks, it wont just be the single block that has a change, but potentially all the others around it. Having the javascript POST all the blocks would go a long way in simplifying that?

File: src/ContactsTabManager.php [hunk 4]
```diff
@@ -150,75 +157,85 @@ public function getTabs(UserInterface $contact) {
...
-          // Apply the mappings.
-          $this->contextHandler->applyContextMapping($block, $contexts);
-        }
-        catch (ContextException $exception) {
-          return FALSE;
+            // Apply the mappings.
+            $this->contextHandler->applyContextMapping($block, $contexts);
+          }
+          catch (ContextException $exception) {
+            return FALSE;
```
Comment (andrewbelcher) [inline src/ContactsTabManager.php:227]: Should this be a `continue;` so we skip this block and proceed with others?

File: src/ContactsTabManager.php [hunk 5]
```diff
@@ -150,75 +157,85 @@ public function getTabs(UserInterface $contact) {
...
 
-    return $block;
+    return $blocks;
   }
 
   /**
    * {@inheritdoc}
    */
-  public function verifyTab(ContactTabInterface $tab, UserInterface $contact, BlockPluginInterface $block = NULL) {
+  public function verifyTab(ContactTabInterface $tab, UserInterface $contact, array $blocks = []) {
```
Comment (andrewbelcher) [inline src/ContactsTabManager.php:255]: Need to update the docblock for `ContactsTabManagerInterface::verifyTab`

File: src/ContactsTabManager.php [hunk 6]
```diff
@@ -44,6 +43,13 @@ class ContactsTabManager implements ContactsTabManagerInterface {
   protected $contextHandler;
 
   /**
+   * The relationship manager.
+   *
+   * @var \Drupal\ctools\Plugin\RelationshipManagerInterface
+   */
+  protected $relationshipManager;
+
+  /**
```
Comment (andrewbelcher) [inline src/ContactsTabManager.php:52]: Good catch, though feel free to commit this directly to 8.x-1.x!

File: src/Controller/DashboardController.php [hunk 7]
```diff
@@ -59,27 +59,39 @@ public function ajaxTab(UserInterface $user, $subpage) {
       'subpage' => $subpage,
     ]);
 
-    $content = [];
+    $content = [
+      '#theme' => 'contacts_dash_tab_content',
+      '#content' => [
```
Comment (andrewbelcher) [inline src/Controller/DashboardController.php]: There's got to be a better way than using `#content` for this, can left/right not just be children that get put in specific regions? I feel it makes the render array very messy...

File: src/Controller/DashboardController.php [hunk 8]
```diff
@@ -59,27 +59,39 @@ public function ajaxTab(UserInterface $user, $subpage) {
...
+        $content['#content'][$block->getConfiguration()['region']][] = $block_content;
+      }
     }
     else {
       drupal_set_message($this->t('Page not found.'), 'warning');
     }
 
     // Prepend the content with system messages.
-    $content['messages'] = [
+    $content['#content']['messages'] = [
```
Comment (andrewbelcher) [inline src/Controller/DashboardController.php]: Is this another defined region?

File: src/Controller/DashboardController.php [hunk 9]
```diff
@@ -93,4 +105,35 @@ public function ajaxTab(UserInterface $user, $subpage) {
...
+   * @param string $region
+   *   The region the block is being moved to.
+   * @param int $weight
+   *   The weight of the block in that region.
+   *
+   * @todo return a http response code.
+   * @return array
+   *   The response commands.
+   */
+  public function moveBlock($tab, $block, $region, $weight) {
```
Comment (andrewbelcher) [inline src/Controller/DashboardController.php]: See comments on the route...

File: src/Controller/DashboardController.php [hunk 10]
```diff
@@ -93,4 +105,35 @@ public function ajaxTab(UserInterface $user, $subpage) {
...
+
+    $block_config = $tab->getBlock($block);
+    $block_config['region'] = $region;
+    $block_config['weight'] = $weight;
+
+    $tab->setBlock($block, $block_config);
+
+    $tab->save();
+
+    return [];
```
Comment (andrewbelcher) [inline src/Controller/DashboardController.php]: I wonder if we want to return the new block/weight or simply a success so the JS can revert changes that fail?

File: src/Entity/ContactTab.php [hunk 11]
```diff
@@ -129,33 +129,56 @@ public function setRelationships(array $relationships) {
...
+  /**
+   * {@inheritdoc}
+   */
+  public function setBlocks(array $blocks) {
+    foreach ($blocks as $block) {
+      if (empty($block['id'])) {
+        throw new \InvalidArgumentException('Missing required ID for block settings.');
+      }
+    }
+    $this->blocks = $blocks;
```
Comment (andrewbelcher) [inline src/Entity/ContactTab.php:193]: Should we make this call `$this->setBlock` or the other way round so we're not duplicating the verification logic?

File: src/Plugin/Block/ContactSummaryTab.php [hunk 12]
```diff
@@ -105,29 +105,6 @@ public function build() {
       $build['#content']['left']['content']['#title'] = $block->label();
     }
```
Comment (andrewbelcher) [inline src/Plugin/Block/ContactSummaryTab.php]: As per comment on the tab config entity, I think this should be gone entirely now...

File: src/Plugin/Block/ContactsDashboardTabs.php [hunk 13]
```diff
@@ -152,26 +152,39 @@ public function buildContent(array &$build) {
...
     }
     else {
       drupal_set_message($this->t('Page not found.'), 'warning');
     }
 
-    $build['content']['messages'] = [
+    $build['content']['#content']['messages'] = [
       '#type' => 'status_messages',
       '#weight' => -99,
     ];
```
Comment (andrewbelcher) [inline src/Plugin/Block/ContactsDashboardTabs.php]: There's a whole load of tidy up needed to this stuff, as we're duplicating loads between here and the controller and the resulting render arrays are a bit messy, but let's not get bogged down in that for now.

Comment (andrewbelcher) [reply src/Plugin/Block/ContactsDashboardTabs.php]: Maybe an `@todo` in both places would be worth doing now...

File: templates/contacts-dash-tab-content.html.twig [hunk 14]
```diff
@@ -0,0 +1,10 @@
+<div {{ attributes.addClass('dash-summary') }}>
+<div class="dash-left">
+    {{ content.left }}
+  </div>
+  <div class="dash-right">
+    {{ content.right }}
+  </div>
+</div>
+
```
Comment (andrewbelcher) [inline templates/contacts-dash-tab-content.html.twig]: Where does `['#content']['messages']` go?

**Review by andrewbelcher — COMMENTED** [review]
Awesome start on this! Looking like we're headed in all the right directions!

File: config/install/contacts.contact_tab.summary.yml [hunk 15]
```diff
@@ -5,11 +5,40 @@ id: summary
...
+    provider: contacts
+    label_display: 'visible'
+    create: crm_notes
+    mode: view_new
+    edit_link: title
+    edit_id: notes
+    region: right
+    weight: 0
+    context_mapping:
+      entity: profile_notes
```
Comment (andrewbelcher) [inline config/install/contacts.contact_tab.summary.yml]: This is missing the organisation profile.

File: config/schema/contact_tab.schema.yml [hunk 16]
```diff
@@ -22,6 +22,9 @@ contacts.contact_tab.*:
...
         label: 'Relationship'
-    block:
-      type: contacts.block_plugin.[id]
-      label: 'Block'
+    blocks:
+      type: sequence
+      label: 'Relationships'
+      sequence:
+        type: contacts.block_plugin.[id]
+        label: 'Blocks'
```
Comment (andrewbelcher) [inline config/schema/contact_tab.schema.yml:39]: This doesn't look right - Relationships is definitely wrong, but we've also got a label at both levels. Is that right and, if so, should the lower one be singular?

File: contacts.routing.yml [hunk 17]
```diff
@@ -28,3 +28,47 @@ contacts.add_org_form:
...
+
+contacts.dash_block_title:
+  path: '/contacts_update_block/{tab}/{block}'
+  defaults:
+    _controller: '\Drupal\contacts\Controller\DashboardController::updateBlockTitle'
+  options:
+    no_cache: TRUE
+  requirements:
+    _access: 'TRUE'
+#    _csrf_token: 'TRUE'
```
Comment (andrewbelcher) [inline contacts.routing.yml]: Having 4 end points for updating a single config I think is a real smell. I think, even stronger than before, that this should be consolidated and we should have a single endpoint for a tab which merges in all the relevant details.

Comment (andrewbelcher) [reply contacts.routing.yml]: I think a consolidated end point will also make it easier to change our mind on both automatic vs explicit saves and storing things in the tempstore...

**Comment by andrewbelcher** [comment]
Loosely related to the endpoints discussion, I think we are going to want the include the following at some point:

- Locking to avoid two people editing (same tab?) at the same time.
- The additional markup and sidebar will want to be loaded on demand rather than on every page load. A couple options are:
  - Edit mode does a full refresh of that tab's markup with all the additional attributes.
  - We make sure we have identifiers on every block all the time, edit mode provides all the additions.

File: src/Controller/DashboardController.php (hunk 7 above)
Comment (andrewbelcher) [reply src/Controller/DashboardController.php]: Let's ignore until we switch to layout plugins...

File: src/Controller/DashboardController.php (hunk 8 above)
Comment (andrewbelcher) [reply src/Controller/DashboardController.php]: Let's deal with this as part of layout plugins, `@todo` for now...

Comment (yanniboi) [reply src/Controller/DashboardController.php]: Will add a @todo

File: src/Entity/ContactTab.php (hunk 11 above)
Comment (andrewbelcher) [reply src/Entity/ContactTab.php:193]: `setBlocks` will reset `$this->blocks` and then call `setBlock` in a `foreach`

File: src/Plugin/Block/ContactSummaryTab.php (hunk 12 above)
Comment (andrewbelcher) [reply src/Plugin/Block/ContactSummaryTab.php]: Whole block can go

Comment (andrewbelcher) [reply src/Plugin/Block/ContactSummaryTab.php]: Whole block can go now

**Pull request closed without merging**